### DISCUSSION TITLE
d2oracle for dagger-graph

### DISCUSSION THREAD

**Discussion by alixander** [dagger/dagger#12523] [Ideas]
Hi @aluzzardi, I saw that you were using D2 for https://github.com/dagger/dagger/tree/main/cmd/dagger-graph, looks great!

I wanted to just bring up that we have an API that can make programmatic generation of D2 diagrams a lot easier/simpler to manage/debug than writing the script with fmt lines: https://pkg.go.dev/oss.terrastruct.com/d2@v0.1.5/d2oracle

Here's a blog post with runnable code on how it might be used: https://terrastruct.com/blog/post/generate-diagrams-programmatically/. Anyway, no action for this issue, just floating it by in case it's helpful and was missed. happy to help if you have any q's!

**Comment by aluzzardi** [comment]
Neat, thanks @alixander! Indeed I missed it -- I went through the `1-d2lib` example and entirely missed the `2-d2oracle` one.

By the way, is there a way using d2oracle to set a label programmatically and skip "interpretation" altogether?

While generating using a string, I ran into escaping problems (single/double quotes,`$` being interpreted etc). The graph "labels" are extracted from logs so they can contain *anything*.

**Comment by alixander** [comment]
@aluzzardi, if you just `Set` (https://pkg.go.dev/oss.terrastruct.com/d2@v0.1.5/d2oracle#Set), D2 will figure out the best way to quote the string, handling escaping for you.

See here for an example: 
https://github.com/terrastruct/d2/blob/master/d2oracle/edit_test.go#L806
It's calling `Set` with a string that has both single quotes and double quotes
